### PR TITLE
RUM-17139: Fix gesture target selection for elevated views

### PR DESCRIPTION
**Pull request by jonathanmos**
### What does this PR do?

Fixes RUM tap actions being attributed to an underlying view when overlapping views use elevation. Tap target selection now compares the first diverging view branches, preferring higher Z and falling back to child index order when Z is equal or non-orderable. Each target is compared using its host view, allowing the same ordering logic to work for mixed Compose and native targets. Existing behavior for nested views and scroll targets is preserved.

Custom child drawing order is not inspected because Android does not publicly expose whether it is enabled for an arbitrary ViewGroup; calling the drawing-order callback unconditionally could disagree with Android’s actual rendering order.

At a high level, when a tap occurs:
- GesturesListener walks the visible view hierarchy and evaluates matching tap targets as they are found.

- Each target is synchronously evaluated against the best target found so far.

- If both targets belong to the same view branch, the deeper target wins.

- If they belong to different branches, their first diverging sibling views determine which branch is visually on top:
Higher Z wins.
Equal or non-orderable Z falls back to the later child index.

- After the hierarchy has been fully traversed, the winning target is used for the RUM tap action.

For Compose targets, the selector uses the Compose host view to perform the same visual-order comparison. Scroll gestures continue using the existing selection behavior.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](../CONTRIBUTING.md) doc)

### CODE REVIEW

**Review by ambushwork — DISMISSED** [review]

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt [hunk 1]
```diff
@@ -146,7 +146,8 @@ internal class GesturesListener(
         // Index 0 is always safe
         @Suppress("UnsafeThirdPartyFunctionCall")
         queue.add(0, decorView)
-        var target: ViewTarget? = null
+        var scrollTarget: ViewTarget? = null
+        val tapTargetSelector = TapTargetSelector()
```
Comment (ambushwork) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt]: `findTarget` will be frequently called, do we need to initialize `TapTargetSelector` every time?

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+import android.os.Build
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.RequiresApi
+import com.datadog.android.internal.system.BuildSdkVersionProvider
+import com.datadog.android.rum.tracking.ViewTarget
+import java.util.IdentityHashMap
+
+internal class TapTargetSelector(
```
Comment (0xnm) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt]: there is quite a lot of iterating through different collections in this class, is there a performance penalty? I see there is `DrawingOrderCache`, but worth to profile the change

Comment (jonathanmos) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt]: I removed DrawingOrderCache because Android doesn’t expose whether custom ordering is enabled, so using it could give incorrect results. The code is simpler now but I’ll still profile it locally.

Comment (jonathanmos) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt]: I had Codex run the benchmark five times. The added overhead was around 0.3 μs for a normal case and 76–79 μs for an extreme case, so it doesn’t look like a performance concern.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt (hunk 1 above)
Comment (jonathanmos) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt]: I profiled it locally. Creating it per call adds negligible overhead (~0.3 μs and 240 bytes for a typical two-candidate lookup), so I’d keep it scoped to each lookup.

Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt]: Does it have any side-effect of moving this allocation to the class level? if not, even if it just saves tiny bit of overhead we should go for it.

**Review by ambushwork — APPROVED** [review]

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Target depth and Z within each branch are irrelevant after the paths diverge. Android
+        // draws the two sibling branches as units, so their ordering decides which target is on top.
+        return currentBranch.isDrawnOnTopOf(candidateBranch, commonParent)
+    }
+
+    private fun View.isDrawnOnTopOf(candidateView: View, parent: ViewGroup): Boolean {
+        return when {
+            z > candidateView.z -> true
+            z < candidateView.z -> false
```
[the remark this answers is not in the document]
Comment (jonathanmos) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/TapTargetSelector.kt:65]: this requires a very unusual NaN Z setup. I’d rather keep this fix focused than add significantly more complexity for that edge case.

**Review by 0xnm — APPROVED** [review]
I'm not sure I follow the full algorithm, but there is a solid test coverage.

I think in case if there are 2 elements with same position and Z1 > Z2, and element 1 is visible, but let's say transparent (maybe some blur layer) and is passing touch events to the element 2, we will report still element 1? Or not?

Although such top layers better done with `Drawable`, probably.

**Comment by jonathanmos** [comment]
If the overlay isn’t clickable, it’s ignored and the underlying view (element 2) is reported. If it’s clickable but forwards touches through custom handling, the overlay (element 1) may be reported instead—but that limitation already existed before this change.
